cerk/src/kernel/broker_event.rs

```diff
@@ -41,26 +53,84 @@ pub enum BrokerEvent {
     /// The event is produced by an input port and is sent to the Kernel. The Kernel sends the same event to the router.
     ///
     /// # Arguments
     ///
     /// * `InternalServerId` - id of the component which received the CloudEvent
+    /// * `CloudEventMessageRoutingId` - the unique identifier of the CloudEvent routing attempt
     /// * `CloudEvent` - the deserialized CloudEvent that the component has received
+    /// * `CloudEventRoutingArgs` - routing arguments to define how a CloudEvent should be routed
+    ///
+    IncomingCloudEvent(
+        InternalServerId,
+        CloudEventMessageRoutingId,
+        CloudEvent,
+        CloudEventRoutingArgs,
+    ),
+
+    /// The `RoutingResult` is the result of a routing from one `IncomingCloudEvent`.
+    /// The event is sent from the router to the kernel and there forwarded as `OutgoingCloudEvent` to the ports.
     ///
-    IncommingCloudEvent(InternalServerId, CloudEvent),
+    /// The `RoutingResult` doesn't have to be sent if there are no destinations (`Vec<OutgoingCloudEvent>.len() == 0`)
+    ///
+    /// # Arguments
+    ///
+    /// * `CloudEventMessageRoutingId` - the unique identifier of the CloudEvent routing attempt
+    /// * `InternalServerId` - the id of the component that received the event from the outside world
+    /// * `Vec<OutgoingCloudEvent>` - the list of events that should be forwarded to the outgoing ports. This vec must only contain `BrokerEvent::OutgoingCloudEvent`. `CloudEventMessageRoutingId` in every message has to be the same as the main `CloudEventMessageRoutingId`.
+    /// * `CloudEventRoutingArgs` - routing arguments to define how a CloudEvent should be routed - this config is used by the kernel; the args for the ports are inside the `Vec<OutgoingCloudEvent>`
+    ///
+    RoutingResult(
+        CloudEventMessageRoutingId,
+        InternalServerId,
+        Vec<BrokerEvent>,
+        CloudEventRoutingArgs,
+    ),
 
     /// The OutgoingCloudEvent event indicates to the receiver that a CloudEvent has been routed and is ready to be forwarded to the outside world.
-    /// The event is created by the router, send to the Kernel and then to the output port.
+    /// The event is created by the router, send to the Kernel (in a badge as `RoutingResult`) and then to the output port(s).
     /// One event for every output port which should forward the data is created.
     ///
     /// # Arguments
     ///
+    /// * `CloudEventMessageRoutingId` - the unique identifier of the CloudEvent routing attempt
     /// * `CloudEvent` - the CloudEvent which should be forwarded
     /// * `InternalServerId` - the id of the component that should send the event
+    /// * `CloudEventRoutingArgs` - routing arguments to define how a CloudEvent should be routed
+    ///
+    OutgoingCloudEvent(
+        CloudEventMessageRoutingId,
+        CloudEvent,
+        InternalServerId,
+        CloudEventRoutingArgs,
+    ),
```

Comment:
would it make sense to put these fields into a struct? I think it would be easier to work with later on (I had to check the docstrings a lot).
```
struct OutgoingCloudEventPayload {
  pub message_routing_id: CloudEventMessageRoutingId,
  pub event: CloudEvent,
  pub source_id: InternalServerId,
  pub routing_args: CloudEventRoutingArgs,
}
```

Reply:
maybe even a single payload struct could be used for all routing events
would it make sense to put these fields into a struct? I think it would be easier to work with later on (I had to check the docstrings a lot).

```
struct EventRoutingPayload {
  pub message_routing_id: CloudEventMessageRoutingId,
  pub event: CloudEvent,
  pub source_id: InternalServerId,
  pub destination_ids: Option<Vec<InternalServerId>>,
  pub routing_args: CloudEventRoutingArgs,
}
```